### ISSUE TITLE
Integrated tmux with Kitty and Sway

### ISSUE THREAD

**Issue by romkatv** [romkatv/zsh4humans#135]
From https://github.com/romkatv/zsh4humans/issues/134 by @maximbaz 

> tmux integration might be another thing that needs considering, but honestly speaking I haven't been using it for a good while now, due to me using too often `Ctrl+Shift+N` to start new terminal in the same folder, and that sway issue where in a freshly new opened window, prompt appears not always in the bottom but in random locations in the middle.

More on the issue with `Ctrl+Shift+N` in Kitty:

- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-798978021
- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-828522833
- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-828554755

More on the issue with Sway:

- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-720134760
- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-720985143
- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-721941970
- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-735213838
- https://github.com/romkatv/zsh4humans/issues/35#issuecomment-735216164

**Comment by romkatv** [comment]
Heh, I should've expected this. That `LINES=100` causes 99 empty lines before the first prompt, just as it should. However, this doesn't help because when the kitty window is consequently resized, prompt stays on lines 16-17 and empty space is added below it. Not all terminals do it this way. Some would keep prompt glued at the bottom and would show more of scrollback on the screen.

Please do the following.

1. Update z4h.
2. Add this line right before `z4h init` in your `.zshrc`:
   ```zsh
   z4h tty-wait --timeout-seconds 1.0 --lines-columns-pattern '^(17 164)'
   ```

This will cause a delay of up to one second when TTY dimensions on startup are 17x164. If the terminal is resized quickly enough, the delay should be small. Plus, this bug doesn't trigger often, so most of the time there should be no delay.

Let me know how this works for you. If you see the bug again, post the same debug info as previously.

**Comment by maximbaz** [comment]
Thanks for the above, interesting, applied.

Minor finding with regards to `propagate-cwd yes`: open a new terminal and press `z4h-cd-back` hotkey, it would go to `/tmp/z4h-tmux-cwd-xxx` folder.

**Comment by romkatv** [comment]
> Minor finding with regards to `propagate-cwd yes`: open a new terminal and press `z4h-cd-back` hotkey, it would go to `/tmp/z4h-tmux-cwd-xxx` folder.

Ack. Will fix a bit later.

**Comment by maximbaz** [comment]
Just now got this one, off by one from my previous pattern 😐 

```
$ tail -2 /tmp/z4h.log && print -Pr -- "[%*] PID=$$, TERM=$TERM, LINES=$LINES, COLUMNS=$COLUMNS"
[14:11:59] PID=1069013, TERM=xterm-kitty, LINES=16, COLUMNS=163
[14:11:59] PID=1069015, TERM=tmux-256color, LINES=16, COLUMNS=163
[14:12:05] PID=1069015, TERM=tmux-256color, LINES=85, COLUMNS=273
```

**Comment by romkatv** [comment]
Let’s extend the pattern: `^(17 164|16 163)`

**Comment by kovidgoyal** [comment]
> Heh, I should've expected this. That `LINES=100` causes 99 empty lines before the first prompt, just as it should. However, this doesn't help because when the kitty window is consequently resized, prompt stays on lines 16-17 and empty space is added below it. Not all terminals do it this way. Some would keep prompt glued at the bottom and would show more of scrollback on the screen.

Note that this is configurable in kitty: https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.scrollback_fill_enlarged_window

**Comment by maximbaz** [comment]
Thanks for the tip! 

@romkatv it just happened again, would you like me to continue adjusting the pattern or try `scrollback_fill_enlarged_window yes` option (with or without `LINES=100`)?

```
$ tail -2 /tmp/z4h.log && print -Pr -- "[%*] PID=$$, TERM=$TERM, LINES=$LINES, COLUMNS=$COLUMNS"
[10:56:02] PID=1710632, TERM=xterm-kitty, LINES=51, COLUMNS=158
[10:56:02] PID=1710634, TERM=tmux-256color, LINES=51, COLUMNS=158
[10:56:06] PID=1710634, TERM=tmux-256color, LINES=85, COLUMNS=273
```

```
$ tail -2 /tmp/z4h.log && print -Pr -- "[%*] PID=$$, TERM=$TERM, LINES=$LINES, COLUMNS=$COLUMNS"
[10:59:00] PID=1713911, TERM=xterm-kitty, LINES=69, COLUMNS=164
[10:59:00] PID=1713913, TERM=tmux-256color, LINES=69, COLUMNS=164
[10:59:02] PID=1713913, TERM=tmux-256color, LINES=85, COLUMNS=273
```

**Comment by maximbaz** [comment]
Reproduces with `scrollback_fill_enlarged_window yes` and `LINES=100` too (with `z4h tty-wait` commented out):

```sh
$ tail -2 /tmp/z4h.log && print -Pr -- "[%*] PID=$$, TERM=$TERM, LINES=$LINES, COLUMNS=$COLUMNS"
[12:30:37] PID=1882756, TERM=xterm-kitty, LINES=39, COLUMNS=143
[12:30:37] PID=1882758, TERM=tmux-256color, LINES=100, COLUMNS=143
[12:30:39] PID=1882758, TERM=tmux-256color, LINES=85, COLUMNS=273
```

I also found a way to reproduce this consistently! Might still be dependent on my usage or specific setup, but at least now we can hopefully iterate faster to find an appropriate solution.

TLDR: the floating windows are causing this problem!

Repro steps for me are:

1. Launch kitty (in my case it by default opens in sway tabbed mode)
    1. Make the window floating
    1. Resize to any size
    1. `echo $LINES` and keep a note of its output (suppose it's `39`)
1. Close the floating kitty
1. Open kitty again (opens again in sway tabbed mode)
    1. At this point the bug will reproduce, and the first line of the three in our `tail -2` output will contain `39` (exactly the output above, in fact).

And I solved the mystery of the magical `17x164` constant 😁  I rarely use floating window in sway, but one purpose I have is to implement a sort of `dmenu` using floating kitty, when I click on a link in terminal, I get a floating kitty window with a selector "what to do with the link - open in browser, copy to clipboard, play in mpv", and that window happens to be `17x164` 🙂 So then once that floating window gets closed, _next time_ I open a kitty window as usual, I'd get this bug.

**Comment by romkatv** [comment]
> Note that this is configurable in kitty: https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.scrollback_fill_enlarged_window

@kovidgoyal Thanks, that's useful to know.

> @romkatv it just happened again, would you like me to continue adjusting the pattern or try `scrollback_fill_enlarged_window yes` option (with or without `LINES=100`)?

I'm fixing this within tmux. If I succeed, `LINES=100` will work in all terminals. It'll also work in kitty regardless of the value of `scrollback_fill_enlarged_window`. I'll let you know when I have something you can test.

**Comment by romkatv** [comment]
> Minor finding with regards to `propagate-cwd yes`: open a new terminal and press `z4h-cd-back` hotkey, it would go to `/tmp/z4h-tmux-cwd-xxx` folder.

This is fixed.

> I'm fixing this within tmux. If I succeed, `LINES=100` will work in all terminals. It'll also work in kitty regardless of the value of `scrollback_fill_enlarged_window`. I'll let you know when I have something you can test.

This ended up too messy. While I made it work well with kitty, on some terminals it results in duplicated scrollback. I don't want to push or to maintain this awfulness. You can remove `LINES=100` from your config.

At this point I see these options:

1. Adjust the value of `--lines-columns-pattern` so that it matches your needs. For example, if your normal TTY is always 85 lines tall, use `85 <->`. Or, if your floating windows are no more than 60 lines tall while your normal TTY is always at least 60 lines tall, use `<60-> <->`. Basically, `"$LINES $COLUMNS"` from normal terminals should match this pattern but the same expression from floating terminals should not. You can test patterns like this:
   ```zsh
   () {
     emulate -L zsh -o extended_glob
     [[ "$LINES $COLUMNS" == $~1 ]] && print match || print no-match
   } '<60-> <->'
   ```
2. Whenever the bug triggers for you, press <kbd>Ctrl+L</kbd>. Low tech but simple.
3. I can try to automate (2) but it's tricky and there will still be blinking when prompt is moved to the bottom.

The first option would be best as long as you can somehow distinguish between normal and floating terminals by their dimensions. Note that neither false positives nor false negatives will result in terrible failures. If a floating terminal matches the pattern, you'll occasionally have prompt in the middle of the screen (which you can move to the bottom with <kbd>Ctrl+L</kbd>). If a normal terminal doesn't match the pattern, you'll have a one second delay before the first prompt. So if you can write a pattern that *almost always* matches normal terminals and *almost never* matches floating terminals, that might be good enough.

**Comment by maximbaz** [comment]
Thanks for looking into this! I agree that this is not critical at all, `Ctrl+L` is definitely an easy workaround, I also wouldn't go with automating it. Now that I know that the root cause is floating windows, I think for my own use-case I can definitely come up with a pattern that will be good enough. 

I think all the items were solved, we can close the ticket. If I'll stumble upon something else (e.g. while using `propagate-cwd`) I'll definitely let you know.

**Comment by romkatv** [comment]
Sounds good. Closing a fixed.
